Agents/PropertyValueEstimationAgent/agent/agentlogic/value_estimation.py

```diff
@@ -0,0 +1,214 @@
+        Prio1: LRPPI:TransactionRecord & OntoBuiltEnv:PropertyPriceIndex
+        Prio2: OntoBuiltEnv:AveragePricePerSqm & OM:Area
+
+        Arguments:
+            transaction_iri {str} - IRI of LRPPI:TransactionRecord
+            prop_price_index_iri {str} - IRI of OntoBuiltEnv:PropertyPriceIndex
+            avgsqm_price_iri {str} - IRI of OntoBuiltEnv:AveragePricePerSqm
+            floor_area_iri {str} - IRI of OM:Area
+        Returns:
+            Graph to instantiate/update property market value
+        """
+
+        # Initialise market value and return triples
+        market_value = None
+        triples = ''
+
+        # Prio 1: Check if transaction record and property price index are provided
+        # (i.e. market value assessment based on previous transaction)
+        if transaction_iri and prop_price_index_iri:
+            # Initialise TS client
+            ts_client = TSClient(kg_client=self.sparql_client)
+            # 1) Retrieve representative UK House Price Index and parse as Series (i.e. unwrap Java data types)
+            # UKHPI was set at a base of 100 in January 2015, and reflects the change in value of residential property since then
+            # (https://landregistry.data.gov.uk/app/ukhpi/doc)
+            try:
+                # Retrieve time series in try-with-resources block to ensure closure of RDB connection
+                with ts_client.connect() as conn:
+                    ts = ts_client.tsclient.getTimeSeries([prop_price_index_iri], conn)
+                dates = [d.toString() for d in ts.getTimes()]
+                values = [v for v in ts.getValues(prop_price_index_iri)]
+            except Exception as ex:
+                self.logger.error('Error retrieving/unwrapping Property Price Index time series')
+                raise TSException('Error retrieving/unwrapping Property Price Index time series') from ex
+
+            # Create UKHPI series with conditioned date index
+            ukhpi = pd.Series(index=dates, data=values)
+            ukhpi = ukhpi.astype(float)
+            ukhpi.index = pd.to_datetime(ukhpi.index, format=TIME_FORMAT_LONG)
+            ukhpi.sort_index(ascending=False, inplace=True)
+            ukhpi.index = ukhpi.index.strftime(TIME_FORMAT_SHORT)
+
+            # 2) Retrieve previous sales transaction details for previous transaction IRI
+            #    and adjust to current market value
+            res = self.sparql_client.get_transaction_details(transaction_iri)
+            ukhpi_now = ukhpi.iloc[0]
+            ukhpi_old = ukhpi[res['date']]
+            market_value = res['price'] * ukhpi_now / ukhpi_old
+
+        # Prio 2: Otherwise assess market value based on FloorArea and AveragePricePerSqm
+        elif avgsqm_price_iri and floor_area_iri and not market_value:
+            # TODO: How to handle cases where avgsqm_price is just not calculated yet?
+            # --> Suggestion: Avoid this case by marking AvgSqmPrice as Synchronous Derivation initially
```

Comment:
As discussed, the `AvgSqmPrice` can still be marked as async derivation and it will work fine if the value is NOT calculated yet. The issue is when the derivation instance IRI is added to the input of `PropertyValueEstimation` derivation directly without knowing the outputs values are already computed. A potential feature request to make this more robust is documented in https://github.com/cambridge-cares/TheWorldAvatar/issues/456

Reply:
Still not fully clear to me. Can you please elaborate which case would work and which one not? Might be easiest to discuss in person, just typing down here for clarification:
```
# AvgSqmPrice derivation
# inputs: property price index, post code, transactions
# outputs: average price (+ respective Measure instance)

# Instantiated input instances
deriv_inputs1 = ['PropertyPriceIndex_1', 'PostalCode_1', 'Transaction_1', 'Transaction_2', 'Transaction_3']
# Mark up derivation and create derivation IRI
derivation_iri1  = derivation_client.createAsyncDerivationForNewInfo(agent1.agentIRI, deriv_inputs1)
--> This will create the derivation_IRI but potentially not the AvgSqmPrice_IRI, correct?

# PropertyValueEstimation derivation
# inputs: property price index, transaction, floor area, AvgSqmPrice
# outputs: AmountOfMoney (+ respective Measure instance)

# Instantiated input instances
deriv_inputs2 = ['PropertyPriceIndex_1', 'Transaction_1', 'Area_1', 'AvgSqmPrice_1']
deriv_inputs3 = ['PropertyPriceIndex_1', 'Transaction_1', 'Area_1', 'derivation_iri1']
# Mark up derivation and create derivation IRI
derivation_iri2  = derivation_client.createAsyncDerivationForNewInfo(agent2.agentIRI, deriv_inputs2)
--> This should work just fine in case AvgSqmPrice_1 has been computed by derivation agent, correct?
--> Otherwise, this IRI shouldn't even exist yet, correct?

derivation_iri3  = derivation_client.createAsyncDerivationForNewInfo(agent2.agentIRI, deriv_inputs3)
--> Would this even work?
```

As discussed, marking `AvgSqmPrice` as synchronous derivation might be most suitable, as this would 
1) Add showcasing `synchronous derivations` to use case
2) Ensure that required inouts for 2nd derivation are available

Reply:
> ```
> # AvgSqmPrice derivation
> # inputs: property price index, post code, transactions
> # outputs: average price (+ respective Measure instance)
> 
> # Instantiated input instances
> deriv_inputs1 = ['PropertyPriceIndex_1', 'PostalCode_1', 'Transaction_1', 'Transaction_2', 'Transaction_3']
> # Mark up derivation and create derivation IRI
> derivation_iri1  = derivation_client.createAsyncDerivationForNewInfo(agent1.agentIRI, deriv_inputs1)
> --> This will create the derivation_IRI but potentially not the AvgSqmPrice_IRI, correct?
> ```

Correct, the `AvgSqmPrice_IRI` will be created by the agent later (depending on the time interval config).

> ```
> # PropertyValueEstimation derivation
> # inputs: property price index, transaction, floor area, AvgSqmPrice
> # outputs: AmountOfMoney (+ respective Measure instance)
> 
> # Instantiated input instances
> deriv_inputs2 = ['PropertyPriceIndex_1', 'Transaction_1', 'Area_1', 'AvgSqmPrice_1']
> deriv_inputs3 = ['PropertyPriceIndex_1', 'Transaction_1', 'Area_1', 'derivation_iri1']
> # Mark up derivation and create derivation IRI
> derivation_iri2  = derivation_client.createAsyncDerivationForNewInfo(agent2.agentIRI, deriv_inputs2)
> --> This should work just fine in case AvgSqmPrice_1 has been computed by derivation agent, correct?
> --> Otherwise, this IRI shouldn't even exist yet, correct?
> ```

Correct, this will work fine.

> ```
> derivation_iri3  = derivation_client.createAsyncDerivationForNewInfo(agent2.agentIRI, deriv_inputs3)
> --> Would this even work?
> ```

This will work also fine, **as long as** `AvgSqmPrice_IRI` does **NOT** exist at the moment you execute this line. The framework only fails when `AvgSqmPrice_IRI` is already computed but `derivation_iri1` was still used as derivation input. Any derivation IRIs passed in the `List<String> inputsAndDerivations` **MUST** be async derivations that have **NOT** got the output computed. For a lengthy job, this is relatively easy to ensure, but for a fast calculation, it will be risky. I hope this is clearer to you?

> As discussed, marking `AvgSqmPrice` as synchronous derivation might be most suitable, as this would
> 
> 1. Add showcasing `synchronous derivations` to use case
> 2. Ensure that required inouts for 2nd derivation are available

Agreed.